engines/e_pkcs11.c

```diff
@@ -0,0 +1,504 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "e_pkcs11.h"
+#include "e_pkcs11_err.c"
+
+static int pkcs11_logout(CK_SESSION_HANDLE session);
+typedef CK_RV pkcs11_pFunc(CK_FUNCTION_LIST **pkcs11_funcs);
+static void pkcs11_finalize(void);
+static void pkcs11_end_session(CK_SESSION_HANDLE session);
+static CK_RV pkcs11_load_functions(const char *library_path);
+static CK_FUNCTION_LIST *pkcs11_funcs;
+
+int pkcs11_rsa_sign(int alg, const unsigned char *md,
+                    unsigned int md_len, unsigned char *sigret,
+                    unsigned int *siglen, const RSA *rsa)
+{
+    CK_RV rv;
+    PKCS11_CTX *ctx;
+    CK_ULONG num;
+    CK_MECHANISM sign_mechanism = { 0 };
+    CK_BBOOL bAwaysAuthentificate = CK_TRUE;
+    CK_ATTRIBUTE keyAttribute[1];
```

Comment:
better init with  `= { 0 };`

Reply:
CK_ATTRIBUTE keyAttribute = { 0 } ?
expected CK_ATTRIBUTE_PTR but argument is of type CK_ATTRIBUTE at row 66

Reply:
I was thinking to this form:
`     CK_ATTRIBUTE keyAttribute[1] = { 0 }; `

Reply:
with { 0 }
warning: missing braces around initializer
CK_ATTRIBUTE keyAttribute[1] = {{ 0 }};  ?